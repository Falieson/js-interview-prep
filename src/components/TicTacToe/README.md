# Tic Tac Toe

## MVP

- [x] Use `cssGrid` to create the game board
- [ ] Determine the winner:
  - [x] across
  - [ ] down
  - [ ] diagonal

## More CSS

- [ ] Create a header w/ `cssGrid` to keep scores for P1 & P2
- [ ] Keep board size symmetrical [ref](https://jsfiddle.net/kwgum8yL/)
- [ ] Center board on the page

## Apollo Cache

- [ ] Use apollo cache to keep the score
- [ ] Create a History page that lists the winner of each game

## Redux

- [ ] Replace apollo w/ redux

## More React

- [ ] Players can choose name
- [ ] Players can choose marker

## Tests

- [ ] Gameplay
  - [ ] One Game
  - [ ] Next Game
  - [ ] History of Games
- [ ] Rendering
