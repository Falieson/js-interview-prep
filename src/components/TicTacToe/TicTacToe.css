/* Helpers that should be moved to a central file once used elsewhere  */
.flex--center {
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

.box--center {
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%);
}

/* TicTacToe css */
.tictactoe--container {
  min-width: 380px;
  min-height: 420px;
}

.tictactoe--board {
  display: grid;
  grid-template-areas: 'board'; /*  scores */
  grid-template-rows:    repeat(3, 120px);
  grid-template-columns: repeat(3, 1fr);
  grid-column-gap: 10px;
  grid-row-gap: 11px;
  border: 1px solid white;
  box-shadow: darkgrey 2px 2px;
}

.tictactoe--cell {
  background-color: lightgrey;
  color: #000;
}
