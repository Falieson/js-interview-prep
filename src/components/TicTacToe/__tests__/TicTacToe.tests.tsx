import { shallow } from 'enzyme'
import * as React from 'react'

import TicTacToe from '../TicTacToe'

describe('<TicTacToe />', () => {
  describe('Helpers', () => {
    describe('until there is a winner', () => {
      it('across', () => {
        const positions = [
          [1, 1, 1],
          [2, 2, 0],
          [0, 0, 0],
        ]
        const TTT = new TicTacToe({positions})

        const result = TTT.checkAcross()
        expect(result.winner).toEqual(1)
      })
      // it('down', () => false)
      // it('diagonally', () => false)
    })
  })
  describe('Plays', () => {
  //   it('with alternating players', () => false)
  //   it('with two players', () => false)
  //   it('until there is a tie', () => false)
    describe('until there is a winner', () => {
    //     it('diagonally', () => false)
    //     it('across', () => false)
    //     it('down', () => false)
    })
  })
  describe('Renders', () => {
    it('Succesfully', () => {
      const wrapper = shallow(<TicTacToe />)
      expect(wrapper).toMatchSnapshot()
    })
    describe('<Board />', () => {
      // it('9 positions', () => false)
      // it('highlight of winning 3 positions in a row', () => false)
      // it('highlight of winning 3 positions in a col', () => false)
      // it('highlight of winning 3 positions in a diag', () => false)
    })
    describe('<Header />', () => {
      // it('displays current player', () => false)
      // it('displays new game button when game is over', () => false)
      // it('score of player1 vs player2', () => false)
    })
    describe('<Overlay />', () => {
      // it('overlay with winning player announce', () => false)
      // it('overlay with no-winner announce', () => false)
    })
  })
})
