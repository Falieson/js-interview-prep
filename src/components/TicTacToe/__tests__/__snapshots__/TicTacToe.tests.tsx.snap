// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicTacToe /> Renders Succesfully 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TicTacToe />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        className="tictactoe--board"
      >
        <div
          className="tictactoe--cell flex--center"
          onClick={[Function]}
        >
          0
        </div>
        <div
          className="tictactoe--cell flex--center"
          onClick={[Function]}
        >
          0
        </div>
        <div
          className="tictactoe--cell flex--center"
          onClick={[Function]}
        >
          0
        </div>
        <div
          className="tictactoe--cell flex--center"
          onClick={[Function]}
        >
          0
        </div>
        <div
          className="tictactoe--cell flex--center"
          onClick={[Function]}
        >
          0
        </div>
        <div
          className="tictactoe--cell flex--center"
          onClick={[Function]}
        >
          0
        </div>
        <div
          className="tictactoe--cell flex--center"
          onClick={[Function]}
        >
          0
        </div>
        <div
          className="tictactoe--cell flex--center"
          onClick={[Function]}
        >
          0
        </div>
        <div
          className="tictactoe--cell flex--center"
          onClick={[Function]}
        >
          0
        </div>
      </div>,
      "className": "tictactoe--container content--center",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          Array [
            <div
              className="tictactoe--cell flex--center"
              onClick={[Function]}
            >
              0
            </div>,
            <div
              className="tictactoe--cell flex--center"
              onClick={[Function]}
            >
              0
            </div>,
            <div
              className="tictactoe--cell flex--center"
              onClick={[Function]}
            >
              0
            </div>,
          ],
          Array [
            <div
              className="tictactoe--cell flex--center"
              onClick={[Function]}
            >
              0
            </div>,
            <div
              className="tictactoe--cell flex--center"
              onClick={[Function]}
            >
              0
            </div>,
            <div
              className="tictactoe--cell flex--center"
              onClick={[Function]}
            >
              0
            </div>,
          ],
          Array [
            <div
              className="tictactoe--cell flex--center"
              onClick={[Function]}
            >
              0
            </div>,
            <div
              className="tictactoe--cell flex--center"
              onClick={[Function]}
            >
              0
            </div>,
            <div
              className="tictactoe--cell flex--center"
              onClick={[Function]}
            >
              0
            </div>,
          ],
        ],
        "className": "tictactoe--board",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0_0",
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "tictactoe--cell flex--center",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "0_1",
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "tictactoe--cell flex--center",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "0_2",
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "tictactoe--cell flex--center",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "1_0",
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "tictactoe--cell flex--center",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "1_1",
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "tictactoe--cell flex--center",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "1_2",
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "tictactoe--cell flex--center",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "2_0",
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "tictactoe--cell flex--center",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "2_1",
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "tictactoe--cell flex--center",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "2_2",
          "nodeType": "host",
          "props": Object {
            "children": 0,
            "className": "tictactoe--cell flex--center",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": 0,
          "type": "div",
        },
      ],
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          className="tictactoe--board"
        >
          <div
            className="tictactoe--cell flex--center"
            onClick={[Function]}
          >
            0
          </div>
          <div
            className="tictactoe--cell flex--center"
            onClick={[Function]}
          >
            0
          </div>
          <div
            className="tictactoe--cell flex--center"
            onClick={[Function]}
          >
            0
          </div>
          <div
            className="tictactoe--cell flex--center"
            onClick={[Function]}
          >
            0
          </div>
          <div
            className="tictactoe--cell flex--center"
            onClick={[Function]}
          >
            0
          </div>
          <div
            className="tictactoe--cell flex--center"
            onClick={[Function]}
          >
            0
          </div>
          <div
            className="tictactoe--cell flex--center"
            onClick={[Function]}
          >
            0
          </div>
          <div
            className="tictactoe--cell flex--center"
            onClick={[Function]}
          >
            0
          </div>
          <div
            className="tictactoe--cell flex--center"
            onClick={[Function]}
          >
            0
          </div>
        </div>,
        "className": "tictactoe--container content--center",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            Array [
              <div
                className="tictactoe--cell flex--center"
                onClick={[Function]}
              >
                0
              </div>,
              <div
                className="tictactoe--cell flex--center"
                onClick={[Function]}
              >
                0
              </div>,
              <div
                className="tictactoe--cell flex--center"
                onClick={[Function]}
              >
                0
              </div>,
            ],
            Array [
              <div
                className="tictactoe--cell flex--center"
                onClick={[Function]}
              >
                0
              </div>,
              <div
                className="tictactoe--cell flex--center"
                onClick={[Function]}
              >
                0
              </div>,
              <div
                className="tictactoe--cell flex--center"
                onClick={[Function]}
              >
                0
              </div>,
            ],
            Array [
              <div
                className="tictactoe--cell flex--center"
                onClick={[Function]}
              >
                0
              </div>,
              <div
                className="tictactoe--cell flex--center"
                onClick={[Function]}
              >
                0
              </div>,
              <div
                className="tictactoe--cell flex--center"
                onClick={[Function]}
              >
                0
              </div>,
            ],
          ],
          "className": "tictactoe--board",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0_0",
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": "tictactoe--cell flex--center",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": 0,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "0_1",
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": "tictactoe--cell flex--center",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": 0,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "0_2",
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": "tictactoe--cell flex--center",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": 0,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "1_0",
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": "tictactoe--cell flex--center",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": 0,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "1_1",
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": "tictactoe--cell flex--center",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": 0,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "1_2",
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": "tictactoe--cell flex--center",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": 0,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "2_0",
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": "tictactoe--cell flex--center",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": 0,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "2_1",
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": "tictactoe--cell flex--center",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": 0,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "2_2",
            "nodeType": "host",
            "props": Object {
              "children": 0,
              "className": "tictactoe--cell flex--center",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": 0,
            "type": "div",
          },
        ],
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
